File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.java
class Solution {

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int n : nums) sum += n;

        if (sum % 2 != 0) return false;

        int target = sum / 2;

        Boolean[][] memo = new Boolean[nums.length][target + 1];
        return dp(nums, 0, target, memo);
    }

    private boolean dp(int[] nums, int i, int target, Boolean[][] memo) {

        if (target == 0) return true;
        if (i >= nums.length || target < 0) return false;

        if (memo[i][target] != null) {
            return memo[i][target];
        }

        boolean result = dp(nums, i + 1, target - nums[i], memo)
                      || dp(nums, i + 1, target, memo);

        memo[i][target] = result;
        return result;
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res=[]
        nums.sort()
        #print(nums)
        left, right = 0, len(nums)-1
        for i in range(len(nums)-2):
            if i>0 and nums[i]==nums[i-1]:
                continue
            left=i+1
            right=len(nums)-1
            #print(str(i)+" "+str(left) + " "+str(right))
            
            target=-nums[i]

            while left<right:
                twosum = nums[left]+nums[right]
                if twosum>target:
                    right-=1
                elif twosum<target:
                    left+=1
                else:
                    #print("Appending: "+str(nums[i])+" "+str(nums[left]) + " "+str(nums[right]))
                    res.append([nums[i],nums[left],nums[right]])

                    while left<right and nums[left]==nums[left+1]:
                        left+=1
                    while left<right and nums[right-1]==nums[right]:
                        right-=1
                    
                    left+=1
                    right-=1
                    
        return res
            
